File: app/src/main/java/de/bikebean/app/ui/drawer/history/HistoryFragment.java
package de.bikebean.app.ui.drawer.history;

import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import de.bikebean.app.MainActivity;
import de.bikebean.app.R;
import de.bikebean.app.db.DatabaseEntity;

public abstract class HistoryFragment extends Fragment {

    protected HistoryViewModel st;

    private HistoryAdapter adapter;

    // UI Elements
    protected RecyclerView recyclerView;
    protected TextView noDataText;
    protected Button positionHistoryButton, batteryHistoryButton;

    protected abstract @NonNull HistoryViewModel getNewStateViewModel();

    protected abstract void setupListeners();

    protected abstract @NonNull HistoryAdapter getNewAdapter(final @NonNull Context ctx);

    @Override
    public void onActivityCreated(final @Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        st = getNewStateViewModel();

        setupListeners();
        initRecyclerView();
    }

    @Override
    public void onResume() {
        super.onResume();

        final @NonNull MainActivity activity = (MainActivity) requireActivity();
        activity.setToolbarScrollEnabled(true);
        activity.resumeToolbarAndBottomSheet();
    }

    private void initRecyclerView() {
        final @NonNull Context ctx = requireContext();

        adapter = getNewAdapter(ctx);
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(ctx));
    }

    protected void setStatesToAdapter(@NonNull List<? extends DatabaseEntity> ls) {
        if (ls.size() != 0) {
            adapter.setStates(ls);
            noDataText.setVisibility(View.GONE);
        } else
            noDataText.setVisibility(View.VISIBLE);
    }

    public void navigateToTab(final @NonNull View v) {
        MainActivity act = (MainActivity) requireActivity();
        if (v.getId() == R.id.positionButton)
            act.navigateTo(R.id.position_action, null);
        else
            act.navigateTo(R.id.battery_action, null);
    }
}

File: app/src/main/java/de/bikebean/app/ui/drawer/status/history/HistoryFragment.kt
package de.bikebean.app.ui.drawer.status.history

import android.content.Context
import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import de.bikebean.app.MainActivity
import de.bikebean.app.R
import de.bikebean.app.db.DatabaseEntity

abstract class HistoryFragment : Fragment(), MainActivity.LimitedBackScope {

    protected var st: HistoryViewModel? = null
    private var historyAdapter: HistoryAdapter? = null

    // UI Elements
    protected var recyclerView: RecyclerView? = null
    protected var noDataText: TextView? = null

    protected abstract val newStateViewModel: HistoryViewModel

    protected abstract fun setupListeners()

    protected abstract fun getNewAdapter(ctx: Context): HistoryAdapter

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        st = newStateViewModel
        setupListeners()
        initRecyclerView()
    }

    override fun onResume() {
        super.onResume()
        (requireActivity() as MainActivity).apply {
            setToolbarScrollEnabled(true)
            resumeToolbarAndBottomSheet()
        }
    }

    override fun onBackPressed(): Boolean {
        (requireActivity() as MainActivity).navigateTo(R.id.map_back_action, null)
        return true
    }

    private fun initRecyclerView() {
        requireContext().let { ctx ->
            historyAdapter = getNewAdapter(ctx)
            recyclerView!!.apply {
                adapter = historyAdapter
                layoutManager = LinearLayoutManager(ctx)
            }
        }
    }

    protected fun setStatesToAdapter(ls: List<DatabaseEntity?>) {
        if (ls.isNotEmpty()) {
            historyAdapter!!.setStates(ls)
            noDataText!!.visibility = View.GONE
        } else noDataText!!.visibility = View.VISIBLE
    }
}